alibi_detect/utils/pytorch/kernels.py

```diff
@@ -20,21 +48,122 @@ def sigma_median(x: torch.Tensor, y: torch.Tensor, dist: torch.Tensor) -> torch.
 
     Returns
     -------
-    The computed bandwidth, `sigma`.
+    The logrithm of the computed bandwidth, `log-sigma`.
     """
     n = min(x.shape[0], y.shape[0])
     n = n if (x[:n] == y[:n]).all() and x.shape == y.shape else 0
     n_median = n + (np.prod(dist.shape) - n) // 2 - 1
     sigma = (.5 * dist.flatten().sort().values[n_median].unsqueeze(dim=-1)) ** .5
-    return sigma
+    return sigma.log()
 
 
-class GaussianRBF(nn.Module):
+class KernelParameter(object):
+    """
+    Parameter class for kernels.
+    """
+
     def __init__(
         self,
-        sigma: Optional[torch.Tensor] = None,
-        init_sigma_fn: Callable = sigma_median,
-        trainable: bool = False
+        value: torch.Tensor = None,
+        init_fn: Optional[Callable] = None,
+        requires_grad: bool = False,
+        requires_init: bool = False
+    ) -> None:
+        super().__init__()
+        self.value = nn.Parameter(value if value is not None else torch.ones(1),
+                                  requires_grad=requires_grad)
+        self.init_fn = init_fn
+        self.requires_init = requires_init
+
+
+class BaseKernel(nn.Module):
```

Comment:
Not sure if we should make this an abstract base class (and the `forward` method an abstract method) or is it too much faff to get the multiple inheritance order right etc.? Same with the `tensorflow` equivalent. @ascillitoe

Reply:
Agree it would be worth thinking about, assuming we're actually expecting users to sub-class `BaseKernel` to create their own custom kernels...

Reply:
Any conclusion on this?

Reply:
IMO it would be more consistent with the rest of the library to do `class BaseKernel(nn.Module, ABC):` and:

```
    @abstractmethod
    def kernel_function(self, x: Union[np.ndarray, torch.Tensor], y: Union[np.ndarray, torch.Tensor],
                        infer_parameter: Optional[bool] = False) -> torch.Tensor:
        pass
```

rather than raise `NotImplementedError`. I believe `ABC` behaves quite nicely as a mix-in class. 

I'm not quite so sure whether we need to make `forward` an abstract method, since it is already defined in `BaseKernel`. Either way, the `cd_create_customised_kernel.ipynb` example is really nice for this, but we will need additional documentation on how to make a custom kernel (and possibly how `feature_axis` etc fit into this).